github_poster/loader/todoist_loader.py

```diff
@@ -0,0 +1,142 @@
+import datetime
+import json
+
+import pandas as pd
+import requests
+
+from github_poster.loader.base_loader import BaseLoader
+
+
+class TodoistLoader(BaseLoader):
+    track_color = "#FFE411"
+    unit = "tasks"
+
+    def __init__(self, from_year, to_year, _type, **kwargs):
+        super().__init__(from_year, to_year, _type)
+        self.from_year = from_year
+        self.to_year = to_year
+        self.todoist_token = kwargs.get("todoist_token", "")
+        # another magic number, try 3 times for calling api
+        self.MAXIMAL_RETRY = 3
+
+    @classmethod
+    def add_loader_arguments(cls, parser, optional):
+        # add argument for loader
+        parser.add_argument(
+            "--todoist_token",
+            dest="todoist_token",
+            type=str,
+            required=optional,
+            help="dev token",
+        )
+
+    # call with token
+    def response(self, url, postdata):
+        # headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/536.11 (KHTML, like Gecko) Chrome/20.0.1132.11 TaoBrowser/2.0 Safari/536.11'}
+        headers = {"Authorization": "Bearer {0}".format(self.todoist_token)}
+        res = requests.post(url=url, data=postdata, headers=headers)
+        resposn = res.json()
+        return resposn
+
+    # call with re-try since todoist api some time get 502
+    def response_with_retry(self, url, postdata, times):
+        # time.sleep(1)
+        try:
+            return self.response(url, postdata)
+        except Exception as e:
+            if times >= self.MAXIMAL_RETRY:
+                print(
+                    f">> Exceed maximal retry {self.MAXIMAL_RETRY}, Raise exception..."
+                )
+                raise (e)  # will stop the program without further handling
+            else:
+                times += 1
+                print(f">> Exception, Retry {times} begins...")
+                return self.response_with_retry(url, postdata, times)
+
+    # call todoist dev api to get activity of completed tasks
+    # refer https://developer.todoist.com/sync/v9/#activity
+    def todoist_completed_activity(self, page, limit, offset):
+        data = {
+            "event_type": "completed",
+            "page": page,
+            "limit": limit,
+            "offset": offset,
+        }
+        url = "https://api.todoist.com/sync/v9/activity/get"
+        re = self.response_with_retry(url, data, 1)
```

Comment:
why 1 here? seems 3?

Reply:
this will always be 1. I should use default parameter and hide the magic number here.

Reply:
cool